### DISCUSSION TITLE
Add warning about `app.add_middleware` behavior (middleware order)

### DISCUSSION THREAD

**Discussion by axd1x8a** [fastapi/fastapi#10366] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from starlette.middleware.authentication import AuthenticationMiddleware
from app.middlewares.oauth_backend import OAuth2Backend

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

app.add_middleware(
    AuthenticationMiddleware,
    backend=OAuth2Backend(),
)
```

### Description

In the [starlette code](https://github.com/encode/starlette/blob/e11a7d8471e7b4d859ac4cc33db2ae5263463a35/starlette/applications.py#L127-L130), we see that `add_middleware` adds a middleware to the beginning of list for some reason, which is very unobvious and leads to errors with middlewares that depend on the order in which they are executed (`CORSMiddleware`). The method itself is used only in `fastapi`, and is considered deprecated (https://github.com/encode/starlette/issues/1490#issuecomment-1347199743) in `starlette` and is never mentioned in their documentation.

In the code example above, this behavior leads to CORS errors on improper authorization because `AuthenticationMiddleware` is executed first and its response prevents execution of all other middlewares.

I think it is worth adding a warning about this in the documentation, or better yet, do the same as in `starlette` and mark this method as deprecated.

### Operating System

Other

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

Python 3.11.1

**Comment by Kludex** [comment]
What you understood:

> [...] is considered deprecated [...]

What I said: 

> [...] There have been discussions to deprecate [...]

Nevertheless, that's old. It's not going to be deprecated. The comment I made was a reply for the previous comment, which is a problem we already addressed.

**Reply by axd1x8a** [reply]
Okay, apparently I misunderstood the connection between those words and the lack of mention of this method in the documentation. Nevertheless, information about how exactly it works should be added to the documentation of fastapi, if you search on discussions you can see that people complain about various errors with middwares, which they were able to fix by changing the order of declaration, but never understood the reason for all this.

**Reply by axd1x8a** [reply]
I'll unmark answer tag for now because the main problem has not been solved yet

**Reply by axd1x8a** [reply]
@Kludex Can you reopen discussion please?

**Comment by YuriiMotov** [comment]
So, as I understand, all we should probably do is to clarify the order of multiple middleware execution in docs.

I already answered such question and opened a PR: https://github.com/fastapi/fastapi/discussions/13679#discussioncomment-13086897
Does this explanation sound clear?
